pirania/files/usr/libexec/rpcd/pirania

```diff
@@ -64,6 +70,56 @@ local function add_many_vouchers(msg)
     printJson(result)
 end
 
+local function add_visitor_voucher(msg)
+    local result = {}
+    local hostnameOutput = shell("cat /proc/sys/kernel/hostname")
+    local hostname = string.gsub(hostnameOutput, "%s+", "")
+    local noteHiph = string.gsub(msg.note, "%s+", "-")
+    local note = string.lower(noteHiph)
+    local secret = msg.secret
+    local epoc = msg.epoc
+    local key = hostname..'-visitor-'..note
+    result.hostname = hostname
+    result.key = key
+    local vouchers = ft.map(
+        function (val)
+            local randomKey = val.key
+            local newKey = key..'-'..randomKey
+            return json.encode(val)
+        end
+    , msg.vouchers)
+    local voucher = io.popen("voucher add_many_vouchers", 'w')
+    result.table = table.concat(vouchers, '\n')
+    -- result.success = assert(voucher:write(table.concat(vouchers, '\n')))
+    printJson(result)
+end
+
+local function add_member_voucher(msg)
+    local hostnameOutput = shell("cat /proc/sys/kernel/hostname")
+    local hostname = string.gsub(hostnameOutput, "%s+", "")
+    local result = {}
+    local noteHiph = string.gsub(msg.note, "%s+", "-")
+    local note = string.lower(noteHiph)
+    local secret = msg.secret
+    -- Should come from a shared-state json
+    local epoc = msg.epoc
+    local key = 'm-'..hostname..'-'..note
+    local output = shell('voucher add_voucher '..key..' '..secret..' '..epoc..' 10 10 1')
```

Comment:
The problem of using `shell` with untrusted input not escaped is that we are exposing the system. I don't know which solition is best and I know that other parts of the code use the shell without escaping but we must remove this usages. 

The prefered way is to use lua directly and do something like call `voucher.add_voucher(key, secret, epoc)` function directly.

Reply:
:+1: